File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    int[][] memo;
    int n;
    public int minFallingPathSum(int[][] matrix) {
        n = matrix.length;
        memo = new int[n][n];
        for (int[] row : memo) {
            java.util.Arrays.fill(row, Integer.MAX_VALUE);
        }
        int ans = Integer.MAX_VALUE;
        for (int c = 0; c < n; c++) {
            ans = Math.min(ans, dfs(matrix, 0, c));
        }
        return ans;
    }
    int dfs(int[][] matrix, int r, int c) {
        if (c < 0 || c == n) {
            return Integer.MAX_VALUE;
        }
        if (r == n) {
            return 0;
        }
        if (memo[r][c] != Integer.MAX_VALUE) {
            return memo[r][c];
        }
        int left = dfs(matrix, r + 1, c - 1);
        int down = dfs(matrix, r + 1, c);
        int right = dfs(matrix, r + 1, c + 1);
        int ans = matrix[r][c] + Math.min(left, Math.min(down, right));
        memo[r][c] = ans;
        return ans;
    }
}

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {

        int n = matrix.size();

        for (int r = 1; r < n; r++) {

            for (int c = 0; c < n; c++) {

                int mid = matrix[r - 1][c];

                int left = (c > 0) ? matrix[r - 1][c - 1] : INT_MAX;

                int right = (c < n - 1) ? matrix[r - 1][c + 1] : INT_MAX;

                matrix[r][c] = matrix[r][c] + min(mid, min(left, right));
            }
        }

        int ans = INT_MAX;

        for (int c = 0; c < n; c++) {
            ans = min(ans, matrix[n - 1][c]);
        }

        return ans;
    }
};